### PR TITLE
[rtslib] restoreconfig: add ability to restore/reload single target or storage_object

### PR DESCRIPTION
**Pull request by pkalever**
Problem:
-------
Right now, if any target or storage_object is failed to load as part of
target service bringup or manual restoreconfig, may be because of an issue from
the backend, then there is no way to reload that single storage_object or target,
we need to reload/restart all the targets/storage_objects present in the node
just to load one single target/storage_object, this interrupts on going
I/O (via this path) for all the volumes hosted within the node.

Solution:
--------
Add a capability to reload single target or storage_object without touching other
targets or storage_objects hosted in the node.

usage:
```
$ targetcli
targetcli shell version 2.1.fb49
Copyright 2011-2013 by Datera, Inc and others
For help on commands, type 'help'.

/> help
[...]

AVAILABLE COMMANDS
  The following commands are available in the current path:

[...]
    - restoreconfig [savefile] [clear_existing] [target] [storage_object]
[...]
```
Ex 1. reload single storage_object
```
$ targetcli restoreconfig /etc/target/saveconfig.json clear_existing \
    storage_object=blockx
```

Ex 2. reload single target
```
$ targetcli restoreconfig /etc/target/saveconfig.json clear_existing \
    target=iqn.2016-12.org.gluster-block:xxx
```

Ex 3. reload a given target and storage_object
```
$ targetcli restoreconfig /etc/target/saveconfig.json clear_existing \
    target=iqn.2016-12.org.gluster-block:xxx storage_object=blockx
```

Rebased on: #130 
Needs: https://github.com/open-iscsi/targetcli-fb/pull/144

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

### CODE REVIEW

**Comment by pkalever** [comment]
@maurizio-lombardi @lxbsz please help review :-)

File: rtslib/root.py [hunk 1]
```diff
@@ -278,26 +278,32 @@ def clear_existing(self, confirm=False):
...
+            if (not storage_object and not target) or (storage_object and so.name == storage_object):
+                so.delete()
 
         # If somehow some hbas still exist (no storage object within?) clean
         # them up too.
-        for hba_dir in glob.glob("%s/core/*_*" % self.configfs_dir):
-            os.rmdir(hba_dir)
+        if not (storage_object or target):
+            for hba_dir in glob.glob("%s/core/*_*" % self.configfs_dir):
+                os.rmdir(hba_dir)
```
Comment (lxbsz) [inline rtslib/root.py]: Here should we also remove the configfs_dir related to the "storage_object" and "target" ?

File: rtslib/root.py [hunk 2]
```diff
@@ -327,30 +332,32 @@ def err_func(err_str):
...
-
-            set_attributes(so_obj, so.get('attributes', {}), so_err_func)
 
-            for alua_tpg in so.get('alua_tpgs', {}):
-               try:
-                   ALUATargetPortGroup.setup(so_obj, alua_tpg, err_func)
-               except RTSLibALUANotSupported:
-                   pass
+            if (not storage_object and not target) or (storage_object and so['name'] == storage_object):
+                try:
```
Comment (lxbsz) [inline rtslib/root.py]: Maybe in this loop if the storage_object != None and once we have found it from the config file, we should break the loop directly no matter after all the restore options are done successfully or failed, since it not allows to have the same 'name' in the same 'plugin' and it makes no sense to continue the rest. right ?

File: rtslib/root.py [hunk 3]
```diff
@@ -366,17 +373,19 @@ def so_err_func(x):
             if 'wwn' not in t:
```
Comment (lxbsz) [inline rtslib/root.py]: And same here in this loop.

File: rtslib/root.py (hunk 1 above)
Comment (pkalever) [reply rtslib/root.py]: @lxbsz this is not needed now, especially for single delete's case, as this is trying to delete stale hba dirs if any.
 
https://github.com/open-iscsi/rtslib-fb/pull/118 will anyway delete them if any stales exists before create starts.

File: rtslib/root.py (hunk 2 above)
Comment (pkalever) [reply rtslib/root.py]: Yeah, good catch, should have break'ed.

File: rtslib/root.py (hunk 1 above)
Comment (lxbsz) [reply rtslib/root.py]: Okay, that's cool.
Thanks.

**Comment by pkalever** [comment]
@lxbsz refreshed now, please take a look.

File: rtslib/root.py [hunk 4]
```diff
@@ -327,30 +336,35 @@ def err_func(err_str):
...
-               try:
-                   ALUATargetPortGroup.setup(so_obj, alua_tpg, err_func)
-               except RTSLibALUANotSupported:
-                   pass
+            if (not storage_object and not target) or (storage_object and so['name'] == storage_object):
+                try:
+                    so_cls = so_mapping[so['plugin']]
+                except KeyError:
+                    err_func("'plugin' not defined or invalid in storageobject %s" % so['name'])
+                    continue
```
Comment (lxbsz) [inline rtslib/root.py]: BTW, how about here when there has failures ? 

Since we have already found it but it fails to continue, IMO there is no need to continue the loop too.

But since the failure branch is rare to happen, I am okay with it and it will up to you.

Thanks.

**Comment by maurizio-lombardi** [comment]
Ok I generally think it's a good and useful pull request.

I also think we need some more comments in the code because the nesting is becoming complex and I personally find it a bit hard to follow all those "if (not storage_object and not target) or (storage_object and so.name == storage_object)" kind of conditions.

File: rtslib/root.py (hunk 1 above)
Comment (pkalever) [reply rtslib/root.py]: @lxbsz yeah, I thought it was a rare case, so left it there :-)

But yeah, since you have pointed it, it won't hurt addressing it. Will do.

And in python there is no goto :) which would have mad life easy.

File: rtslib/root.py (hunk 4 above)
Comment (pkalever) [reply rtslib/root.py]: @lxbsz yeah, I thought it was a rare case, so left it there :-)

But yeah, since you have pointed it, it won't hurt addressing it. Will do.

And in python there is no goto :) which would have mad life easy.

**Comment by pkalever** [comment]
@maurizio-lombardi @lxbsz addressed the review, please take a look. Thanks!

**Comment by maurizio-lombardi** [comment]
Good, with the comments it's clearer now what the code does, thanks.

**Comment by lxbsz** [comment]
I am testing it, not done yet.
Thanks.

**Comment by lxbsz** [comment]
It works for me.

```
[root@rhel3 rtslib-fb]# targetcli clearconfig confirm=True
All configuration cleared
[root@rhel3 rtslib-fb]# targetcli ls
o- / ......................................................................................................................... [...]
  o- backstores .............................................................................................................. [...]
  | o- block .................................................................................................. [Storage Objects: 0]
  | o- fileio ................................................................................................. [Storage Objects: 0]
  | o- pscsi .................................................................................................. [Storage Objects: 0]
  | o- ramdisk ................................................................................................ [Storage Objects: 0]
  | o- user:fbo ............................................................................................... [Storage Objects: 0]
  | o- user:file .............................................................................................. [Storage Objects: 0]
  | o- user:glfs .............................................................................................. [Storage Objects: 0]
  | o- user:qcow .............................................................................................. [Storage Objects: 0]
  | o- user:zbc ............................................................................................... [Storage Objects: 0]
  o- iscsi ............................................................................................................ [Targets: 0]
  o- loopback ......................................................................................................... [Targets: 0]
  o- vhost ............................................................................................................ [Targets: 0]
  o- xen-pvscsi ....................................................................................................... [Targets: 0]
[root@rhel3 rtslib-fb]# cat /etc/target/saveconfig.json 
{
  "storage_objects": [
    {
      "alua_tpgs": [
        {
          "alua_access_state": 1, 
          "alua_access_status": 2, 
          "alua_access_type": 1, 
          "alua_support_active_nonoptimized": 1, 
          "alua_support_active_optimized": 1, 
          "alua_support_offline": 0, 
          "alua_support_standby": 0, 
          "alua_support_transitioning": 1, 
          "alua_support_unavailable": 0, 
          "alua_write_metadata": 0, 
          "implicit_trans_secs": 0, 
          "name": "glfs_tg_pt_gp_ano", 
          "nonop_delay_msecs": 100, 
          "preferred": 0, 
          "tg_pt_gp_id": 2, 
          "trans_delay_msecs": 0
        }, 
        {
          "alua_access_state": 0, 
          "alua_access_status": 0, 
          "alua_access_type": 1, 
          "alua_support_active_nonoptimized": 1, 
          "alua_support_active_optimized": 1, 
          "alua_support_offline": 0, 
          "alua_support_standby": 0, 
          "alua_support_transitioning": 1, 
          "alua_support_unavailable": 0, 
          "alua_write_metadata": 0, 
          "implicit_trans_secs": 0, 
          "name": "glfs_tg_pt_gp_ao", 
          "nonop_delay_msecs": 100, 
          "preferred": 0, 
          "tg_pt_gp_id": 1, 
          "trans_delay_msecs": 0
        }, 
        {
          "alua_access_state": 0, 
          "alua_access_status": 0, 
          "alua_access_type": 3, 
          "alua_support_active_nonoptimized": 1, 
          "alua_support_active_optimized": 1, 
          "alua_support_offline": 1, 
          "alua_support_standby": 1, 
          "alua_support_transitioning": 1, 
          "alua_support_unavailable": 1, 
          "alua_write_metadata": 0, 
          "implicit_trans_secs": 0, 
          "name": "default_tg_pt_gp", 
          "nonop_delay_msecs": 100, 
          "preferred": 0, 
          "tg_pt_gp_id": 0, 
          "trans_delay_msecs": 0
        }
      ], 
      "attributes": {
        "cmd_time_out": 130, 
        "dev_config": "glfs/repvol@localhost/block-store/5bf22bfa-965a-48c3-8a43-f0d5018e9303", 
        "dev_size": 104857600, 
        "emulate_write_cache": 0, 
        "nl_reply_supported": 1, 
        "qfull_time_out": -1
      }, 
      "config": "glfs/repvol@localhost/block-store/5bf22bfa-965a-48c3-8a43-f0d5018e9303", 
      "control": "max_data_area_mb=1024,hw_block_size=512", 
      "hw_max_sectors": 128, 
      "name": "block0", 
      "plugin": "user", 
      "size": 104857600, 
      "wwn": "5bf22bfa-965a-48c3-8a43-f0d5018e9303"
    }
  ], 
  "targets": [
    {
      "fabric": "iscsi", 
      "tpgs": [
        {
          "attributes": {
            "authentication": 0, 
            "cache_dynamic_acls": 1, 
            "default_cmdsn_depth": 64, 
            "default_erl": 0, 
            "demo_mode_discovery": 1, 
            "demo_mode_write_protect": 0, 
            "fabric_prot_type": 0, 
            "generate_node_acls": 1, 
            "login_keys_workaround": 1, 
            "login_timeout": 15, 
            "netif_timeout": 2, 
            "prod_mode_write_protect": 0, 
            "t10_pi": 0, 
            "tpg_enabled_sendtargets": 1
          }, 
          "enable": true, 
          "luns": [
            {
              "alias": "48b7728b8a", 
              "alua_tg_pt_gp_name": "glfs_tg_pt_gp_ano", 
              "index": 0, 
              "storage_object": "/backstores/user/block0"
            }
          ], 
          "node_acls": [], 
          "parameters": {
            "AuthMethod": "CHAP,None", 
            "DataDigest": "CRC32C,None", 
            "DataPDUInOrder": "Yes", 
            "DataSequenceInOrder": "Yes", 
            "DefaultTime2Retain": "20", 
            "DefaultTime2Wait": "2", 
            "ErrorRecoveryLevel": "0", 
            "FirstBurstLength": "65536", 
            "HeaderDigest": "CRC32C,None", 
            "IFMarkInt": "Reject", 
            "IFMarker": "No", 
            "ImmediateData": "Yes", 
            "InitialR2T": "Yes", 
            "MaxBurstLength": "262144", 
            "MaxConnections": "1", 
            "MaxOutstandingR2T": "1", 
            "MaxRecvDataSegmentLength": "8192", 
            "MaxXmitDataSegmentLength": "262144", 
            "OFMarkInt": "Reject", 
            "OFMarker": "No", 
            "TargetAlias": "LIO Target"
          }, 
          "portals": [
            {
              "ip_address": "192.168.195.164", 
              "iser": false, 
              "offload": false, 
              "port": 3260
            }
          ], 
          "tag": 3
        }, 
        {
          "attributes": {
            "authentication": 0, 
            "cache_dynamic_acls": 1, 
            "default_cmdsn_depth": 64, 
            "default_erl": 0, 
            "demo_mode_discovery": 1, 
            "demo_mode_write_protect": 0, 
            "fabric_prot_type": 0, 
            "generate_node_acls": 1, 
            "login_keys_workaround": 1, 
            "login_timeout": 15, 
            "netif_timeout": 2, 
            "prod_mode_write_protect": 0, 
            "t10_pi": 0, 
            "tpg_enabled_sendtargets": 0
          }, 
          "enable": false, 
          "luns": [
            {
              "alias": "f83b3796a4", 
              "alua_tg_pt_gp_name": "glfs_tg_pt_gp_ano", 
              "index": 0, 
              "storage_object": "/backstores/user/block0"
            }
          ], 
          "node_acls": [], 
          "parameters": {
            "AuthMethod": "CHAP,None", 
            "DataDigest": "CRC32C,None", 
            "DataPDUInOrder": "Yes", 
            "DataSequenceInOrder": "Yes", 
            "DefaultTime2Retain": "20", 
            "DefaultTime2Wait": "2", 
            "ErrorRecoveryLevel": "0", 
            "FirstBurstLength": "65536", 
            "HeaderDigest": "CRC32C,None", 
            "IFMarkInt": "Reject", 
            "IFMarker": "No", 
            "ImmediateData": "Yes", 
            "InitialR2T": "Yes", 
            "MaxBurstLength": "262144", 
            "MaxConnections": "1", 
            "MaxOutstandingR2T": "1", 
            "MaxRecvDataSegmentLength": "8192", 
            "MaxXmitDataSegmentLength": "262144", 
            "OFMarkInt": "Reject", 
            "OFMarker": "No", 
            "TargetAlias": "LIO Target"
          }, 
          "portals": [
            {
              "ip_address": "192.168.195.163", 
              "iser": false, 
              "offload": false, 
              "port": 3260
            }
          ], 
          "tag": 2
        }, 
        {
          "attributes": {
            "authentication": 0, 
            "cache_dynamic_acls": 1, 
            "default_cmdsn_depth": 64, 
            "default_erl": 0, 
            "demo_mode_discovery": 1, 
            "demo_mode_write_protect": 0, 
            "fabric_prot_type": 0, 
            "generate_node_acls": 1, 
            "login_keys_workaround": 1, 
            "login_timeout": 15, 
            "netif_timeout": 2, 
            "prod_mode_write_protect": 0, 
            "t10_pi": 0, 
            "tpg_enabled_sendtargets": 0
          }, 
          "enable": false, 
          "luns": [
            {
              "alias": "fb9a5ff97c", 
              "alua_tg_pt_gp_name": "glfs_tg_pt_gp_ao", 
              "index": 0, 
              "storage_object": "/backstores/user/block0"
            }
          ], 
          "node_acls": [], 
          "parameters": {
            "AuthMethod": "CHAP,None", 
            "DataDigest": "CRC32C,None", 
            "DataPDUInOrder": "Yes", 
            "DataSequenceInOrder": "Yes", 
            "DefaultTime2Retain": "20", 
            "DefaultTime2Wait": "2", 
            "ErrorRecoveryLevel": "0", 
            "FirstBurstLength": "65536", 
            "HeaderDigest": "CRC32C,None", 
            "IFMarkInt": "Reject", 
            "IFMarker": "No", 
            "ImmediateData": "Yes", 
            "InitialR2T": "Yes", 
            "MaxBurstLength": "262144", 
            "MaxConnections": "1", 
            "MaxOutstandingR2T": "1", 
            "MaxRecvDataSegmentLength": "8192", 
            "MaxXmitDataSegmentLength": "262144", 
            "OFMarkInt": "Reject", 
            "OFMarker": "No", 
            "TargetAlias": "LIO Target"
          }, 
          "portals": [
            {
              "ip_address": "192.168.195.162", 
              "iser": false, 
              "offload": false, 
              "port": 3260
            }
          ], 
          "tag": 1
        }
      ], 
      "wwn": "iqn.2016-12.org.gluster-block:5bf22bfa-965a-48c3-8a43-f0d5018e9303"
    }
  ]
}
[root@rhel3 rtslib-fb]#  targetcli restoreconfig /etc/target/saveconfig.json clear_existing storage_object=block0
Configuration restored, 2 recoverable errors:
Storage Object user/block0: Cannot set attribute dev_config: [Errno 95] Operation not supported, skipped
Storage Object user/block0: Cannot set attribute emulate_write_cache: [Errno 95] Operation not supported, skipped
[root@rhel3 rtslib-fb]# targetcli restoreconfig /etc/target/saveconfig.json clear_existing \
>     target=iqn.2016-12.org.gluster-block:5bf22bfa-965a-48c3-8a43-f0d5018e9303
Configuration restored from /etc/target/saveconfig.json
[root@rhel3 rtslib-fb]# targetcli ls
o- / ......................................................................................................................... [...]
  o- backstores .............................................................................................................. [...]
  | o- block .................................................................................................. [Storage Objects: 0]
  | o- fileio ................................................................................................. [Storage Objects: 0]
  | o- pscsi .................................................................................................. [Storage Objects: 0]
  | o- ramdisk ................................................................................................ [Storage Objects: 0]
  | o- user:fbo ............................................................................................... [Storage Objects: 0]
  | o- user:file .............................................................................................. [Storage Objects: 0]
  | o- user:glfs .............................................................................................. [Storage Objects: 1]
  | | o- block0 ........................... [repvol@localhost/block-store/5bf22bfa-965a-48c3-8a43-f0d5018e9303 (100.0MiB) activated]
  | |   o- alua ................................................................................................... [ALUA Groups: 3]
  | |     o- default_tg_pt_gp ....................................................................... [ALUA state: Active/optimized]
  | |     o- glfs_tg_pt_gp_ano .................................................................. [ALUA state: Active/non-optimized]
  | |     o- glfs_tg_pt_gp_ao ....................................................................... [ALUA state: Active/optimized]
  | o- user:qcow .............................................................................................. [Storage Objects: 0]
  | o- user:zbc ............................................................................................... [Storage Objects: 0]
  o- iscsi ............................................................................................................ [Targets: 1]
  | o- iqn.2016-12.org.gluster-block:5bf22bfa-965a-48c3-8a43-f0d5018e9303 ................................................ [TPGs: 3]
  |   o- tpg1 ........................................................................................................... [disabled]
  |   | o- acls .......................................................................................................... [ACLs: 0]
  |   | o- luns .......................................................................................................... [LUNs: 1]
  |   | | o- lun0 ................................................................................. [user/block0 (glfs_tg_pt_gp_ao)]
  |   | o- portals .................................................................................................... [Portals: 1]
  |   |   o- 192.168.195.162:3260 ............................................................................................. [OK]
  |   o- tpg2 ........................................................................................................... [disabled]
  |   | o- acls .......................................................................................................... [ACLs: 0]
  |   | o- luns .......................................................................................................... [LUNs: 1]
  |   | | o- lun0 ................................................................................ [user/block0 (glfs_tg_pt_gp_ano)]
  |   | o- portals .................................................................................................... [Portals: 1]
  |   |   o- 192.168.195.163:3260 ............................................................................................. [OK]
  |   o- tpg3 .................................................................................................. [gen-acls, no-auth]
  |     o- acls .......................................................................................................... [ACLs: 0]
  |     o- luns .......................................................................................................... [LUNs: 1]
  |     | o- lun0 ................................................................................ [user/block0 (glfs_tg_pt_gp_ano)]
  |     o- portals .................................................................................................... [Portals: 1]
  |       o- 192.168.195.164:3260 ............................................................................................. [OK]
  o- loopback ......................................................................................................... [Targets: 0]
  o- vhost ............................................................................................................ [Targets: 0]
  o- xen-pvscsi ....................................................................................................... [Targets: 0]
[root@rhel3 rtslib-fb]# 

```

**Comment by lxbsz** [comment]
Additional tests:

```
[root@rhel3 rtslib-fb]# targetcli clearconfig confirm=True
All configuration cleared
[root@rhel3 rtslib-fb]#  targetcli restoreconfig /etc/target/saveconfig.json clear_existing \
>     target=iqn.2016-12.org.gluster-block:5bf22bfa-965a-48c3-8a43-f0d5018e9303 storage_object=block0
Configuration restored, 2 recoverable errors:
Storage Object user/block0: Cannot set attribute dev_config: [Errno 95] Operation not supported, skipped
Storage Object user/block0: Cannot set attribute emulate_write_cache: [Errno 95] Operation not supported, skipped
[root@rhel3 rtslib-fb]# targetcli ls
o- / ......................................................................................................................... [...]
  o- backstores .............................................................................................................. [...]
  | o- block .................................................................................................. [Storage Objects: 0]
  | o- fileio ................................................................................................. [Storage Objects: 0]
  | o- pscsi .................................................................................................. [Storage Objects: 0]
  | o- ramdisk ................................................................................................ [Storage Objects: 0]
  | o- user:fbo ............................................................................................... [Storage Objects: 0]
  | o- user:file .............................................................................................. [Storage Objects: 0]
  | o- user:glfs .............................................................................................. [Storage Objects: 1]
  | | o- block0 ........................... [repvol@localhost/block-store/5bf22bfa-965a-48c3-8a43-f0d5018e9303 (100.0MiB) activated]
  | |   o- alua ................................................................................................... [ALUA Groups: 3]
  | |     o- default_tg_pt_gp ....................................................................... [ALUA state: Active/optimized]
  | |     o- glfs_tg_pt_gp_ano .................................................................. [ALUA state: Active/non-optimized]
  | |     o- glfs_tg_pt_gp_ao ....................................................................... [ALUA state: Active/optimized]
  | o- user:qcow .............................................................................................. [Storage Objects: 0]
  | o- user:zbc ............................................................................................... [Storage Objects: 0]
  o- iscsi ............................................................................................................ [Targets: 1]
  | o- iqn.2016-12.org.gluster-block:5bf22bfa-965a-48c3-8a43-f0d5018e9303 ................................................ [TPGs: 3]
  |   o- tpg1 ........................................................................................................... [disabled]
  |   | o- acls .......................................................................................................... [ACLs: 0]
  |   | o- luns .......................................................................................................... [LUNs: 1]
  |   | | o- lun0 ................................................................................. [user/block0 (glfs_tg_pt_gp_ao)]
  |   | o- portals .................................................................................................... [Portals: 1]
  |   |   o- 192.168.195.162:3260 ............................................................................................. [OK]
  |   o- tpg2 ........................................................................................................... [disabled]
  |   | o- acls .......................................................................................................... [ACLs: 0]
  |   | o- luns .......................................................................................................... [LUNs: 1]
  |   | | o- lun0 ................................................................................ [user/block0 (glfs_tg_pt_gp_ano)]
  |   | o- portals .................................................................................................... [Portals: 1]
  |   |   o- 192.168.195.163:3260 ............................................................................................. [OK]
  |   o- tpg3 .................................................................................................. [gen-acls, no-auth]
  |     o- acls .......................................................................................................... [ACLs: 0]
  |     o- luns .......................................................................................................... [LUNs: 1]
  |     | o- lun0 ................................................................................ [user/block0 (glfs_tg_pt_gp_ano)]
  |     o- portals .................................................................................................... [Portals: 1]
  |       o- 192.168.195.164:3260 ............................................................................................. [OK]
  o- loopback ......................................................................................................... [Targets: 0]
  o- vhost ............................................................................................................ [Targets: 0]
  o- xen-pvscsi ....................................................................................................... [Targets: 0]
[root@rhel3 rtslib-fb]#
```

**Comment by maurizio-lombardi** [comment]
@lxbsz Thank you very much Xiubo for testing this
